docs/the-git-based-process-you-need-for-mainframe-development-DRAFT/the-git-based-development-process-for-mainframe-development.md

```diff
@@ -0,0 +1,303 @@
+
+All these scenarios lead to the requirement on the development process to implement changes independently from the main workflow (see 2.3.1).
+
+Note that the *epic* branch workflow described in this section is not meant to be used for a single, small feature that a developer wants to hold back for an upcoming release. In those smaller cases, the developer retains the feature branch until the change is planned to be released.
+
+![Figure 11 - Use epic branches for significant development initiatives](images/media/image12.png)
+
+The development tasks for a development initiative are:
+
+1.  The team branches off an *epic* branch, e.g. *epic/epic1234*, from the Git tag representing the current production version of the application, which is at this point the most stable configuration. This process provides them isolation of the codebase from any other ongoing changes for the next iteration(s).
+
+2.  Based on how the work items are distributed between the developers, a feature branch is created according to pre-defined naming conventions such as *epic/epic1234/feature4*, *epic/epic1234/feature5* based on the *epic/epic1234* branch.
+
+![](images/media/image13.png) 
+
+3.  The developers fetch the feature branch from the central Git repository into their local clone of the repository and start making the necessary modifications. They leverage the user build facility of their IDE for building and testing individual programs. They can also leverage *a feature branch pipeline* to build the changed and impacted files. Optionally, the developer can prepare a preliminary package, which can be used for validating the fix in a controlled test environment, such as a EPIC-1-FEATURE-TEST environment[^14].
+
+4.  The developer initiates the Pull Request process, which provides the ability to add peer review and approval steps before allowing the changes to be merged into the *epic* branch.
+
+5.  A *build pipeline*[^15] for the epic branch will build all the merged features (changes and their impacts) from the point the epic branch was branched off.
+
+6.  It is mandatory, that the team is frequently incorporating updates which got implemented for the next release or got released to production via the standard development process via the *main* branch into the *epic* branch to avoid that the configurations diverge too much and make the planned merge hard. A common practice is to at least integrate changes after each completion of a release via the main workflow (See Figure 15) to merge the stable versions, while more frequent integrations may lead to pull intermediate versions of features, which may contain defects.
```

Comment:
> while more frequent integrations may lead to pull intermediate versions of features, which may contain defects.

@dennis-behm @M-DLB If the dev team is pulling from `main` into their epic branch, shouldn't `main` always be buildable/deployable and not contain intermediate or defective versions of features? (Or if they do, shouldn't those intermediate features be behind a feature flag?)

Reply:
We're definitely saying elsewhere that PRs with commits that break the build of `main` should not EVER be merged. 

Beyond that the question becomes more subtle - how many automated tests are part of the build? If the feature is new, is it entirely controlled by a feature-flag until the planned level of function is properly tested (and then the automated tests are updated to include looking for regressions in it).

I don't think `main` needs to be **always** deployable. We have release branches to be the means to fix the last production deployment. Certainly the next release candidate has to be proven to be good enough to deploy, but in between release candidates I think it's realistic to have some in-complete features.

Reply:
Yep, the `main` branch should stay in a state where it can be built fine. If the pipeline build for a feature branch is broken, then this feature branch shouldn't be merged into `main`, and it's the same with hot-fix branches. If there are conflicts to be solved, it should be done in the feature branch and the pipeline for the feature branch should complete, prior to the merge of the feature branch into `main`.
As Ian pointed, a correct pipeline execution for `main` doesn't mean that the packaged artifacts will be deployed somewhere (probably it can be deployed to the integration environment automatically, but that's not the case for other test environments or production-like environments).

Reply:
I suppose it might depend who you ask, whether or not `main` should always be deployable (although I think the general consensus is that it should always at least be buildable).

Maybe there is some ambiguity in what is meant here by "intermediate versions of features". In my mind, this could be an intermediate deliverable as a normal part of incremental, iterative development for that feature. However, it should not be anything that would prevent other teammates from regularly (and frequently) pulling updates from the team's shared branch of code (e.g. `main`) into their own topic branch for fear of pulling in defects.

Reply:
@lauren-li glad that you found this. Indeed, this is misleading:
> while more frequent integrations may lead to pull intermediate versions of features, which may contain defects.

Even that `main` is buildable, the state of `main` might contain code that implements feature that are not yet not functioning (maybe even not tested from a business perspective) and are not yet fully implemented. 

So instead of
> intermediate versions of features, which may contain defects

What about

> while more frequent integrations may lead to pull intermediate versions of features, which may not be fully implemented from a business perspective.

Reply:
>  A common practice is to at least integrate changes after each completion of a release via the main workflow (See Figure 15) to merge the stable versions, while more frequent integrations may lead to pulling intermediate versions of features, which may not be fully implemented from a business perspective.

@dennis-behm I do like this better, but it still seems to imply to me that the developers should be cautious about pulling from the `main` branch "too frequently" (i.e. between releases). In my experience, pulling from such a central shared branch of code between releases should not feel so risky. (Even pulling on a daily basis should be fine.) How about:

> A common practice is to at least integrate changes after each completion of a release via the main workflow (See Figure 15) to merge the stable versions. More frequent integrations may lead to pulling intermediate versions of features, which may not be fully implemented from a business perspective; however, this should not deter developers since the `main` branch should always be in a buildable state. 

Also, I just realized that we seem to be referencing a figure that has no caption (visible figure number) in the `.md` document. We could either re-add captions using HTML, or just reword the text to reference something to the effect of "the following image" for our figures. (We might have some broken references to the draft PDF's section numbers, as well.)

Reply:
@lauren-li thank you for improving my proposal. Please see the update being included in  - https://github.com/IBM/dbb-zappbuild/pull/420